kotlinx-coroutines-core/jvm/test/lincheck/LockFreeListMultiLincheckTest.kt

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2016-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+@file:Suppress("unused")
+
+package kotlinx.coroutines.lincheck
+
+import kotlinx.atomicfu.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.internal.*
+import org.jetbrains.kotlinx.lincheck.annotations.*
+import org.jetbrains.kotlinx.lincheck.paramgen.*
+import org.jetbrains.kotlinx.lincheck.strategy.managed.modelchecking.*
+import kotlin.test.*
+
+@Param.Params(
+    Param(name = "id", gen = IntGen::class, conf = "0:1"),
+    Param(name = "value", gen = IntGen::class, conf = "1:5")
+)
+class LockFreeList2LincheckTest : LockFreeListMultiLincheckTest(2)
+
+@Param.Params(
+    Param(name = "id", gen = IntGen::class, conf = "0:2"),
+    Param(name = "value", gen = IntGen::class, conf = "1:5")
+)
+class LockFreeList3LincheckTest : LockFreeListMultiLincheckTest(3)
+
+/**
+ * Test atomic operations across multiple instances of `LockFreeLinkedList`.
+ */
+abstract class LockFreeListMultiLincheckTest(val n: Int) : AbstractLincheckTest() {
+    private val _opSequence = atomic(0L)
+
+    class Node(val value: Int): LockFreeLinkedListNode() {
+        override fun toString(): String = "N$value"
+    }
+    private val q = Array(n) { index ->
+        object : LockFreeLinkedListHead() {
+            override fun toString(): String = "H$index"
+        }
+    }
+
+    @Operation
+    fun addLast(@Param(name = "id") id: Int, @Param(name = "value") value: Int) {
+        q[id].addLast(Node(value))
+    }
+
+    @Operation
+    fun addTwo(@Param(name = "id") id: Int, @Param(name = "value") value: Int) {
+        var result: Any?
+        do {
+            val add1 = q[id].describeAddLast(Node(value))
```

Comment:
You always follow the same order here ("0, 1", "1, 2", "2, 3") and never the opposite. However, everything is fine with two lists. I do not think that you should fix this, but want to draw some attention.

Reply:
The order is "0,1", "1, 2", and "2,0" (mind the modulo operation), so there is no hierarchy and it properly stresses the deadlock scenario. Moreover, there is variant of the test with n=2, with does both "0,1" and "1,0".  You can comment out `override val opSequence` to disable deadlock-avoidance and check that this test immediately produces `StackOverflow`.

Reply:
👍